feat: index records with suggestions for search engine

frascuchon opened the pull request:
<!-- Thanks for your contribution! As part of our Community Growers initiative 🌱, we're donating Justdiggit bunds in your name to reforest sub-Saharan Africa. To claim your Community Growers certificate, please contact David Berenstein in our Slack community or fill in this form https://tally.so/r/n9XrxK once your PR has been merged. -->

# Description

This PR adds support for indexing suggestions when indexing records in the search index.

In order to be sure that all record attributes are passed when indexing records, a workaround has been implemented by forcing a record, responses, and suggestions to refresh before indexing them. Otherwise is hard to add that info when loading records in the current code base. 

This behavior must be reviewed and simplified cc @gabrielmbmb @jfcalvo 

Changes related to tests will be moved into a separate PR since extra refactoring work must be done. Once this extra PR is merged here, the PR will be marked as ready for review. Related PR https://github.com/argilla-io/argilla/pull/4318


Refs: #3849 

Closes https://github.com/argilla-io/argilla/issues/4230

frascuchon committed:
chore: define property for question type

frascuchon committed:
feat: Adding record suggestions when indexing records

Also, simplify and remove all extra mapper componets and reuse some code blocks
regarding the elastic field definitions

frascuchon committed:
chore: remove unused components

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

frascuchon committed:
fix: Enable required attributes when indexing records

jfcalvo commented on src/argilla/server/contexts/datasets.py:
```diff
@@ -788,6 +788,19 @@ async def _update_record(
+        for response in record.responses:
+            await response.awiatable_attrs.user
+        for suggestion in record.suggestions:
+            await suggestion.awaitable_attrs.question
```
I think we can improve the naming in the function and separate it having one for the collection and one for a single item:

```suggestion
async def _preload_records_associations(records: List[Record]) -> None:
    for record in records:
        await _preload_record_associations(record)


async def _preload_record_associations(record: Record) -> None:
    # TODO: This behavior must be reviewed !!!
    await record.awaitable_attrs.suggestions
    await record.awaitable_attrs.responses
    await record.awaitable_attrs.vectors

    for response in record.responses:
        await response.awiatable_attrs.user

    for suggestion in record.suggestions:
        await suggestion.awaitable_attrs.question


```

jfcalvo commented on src/argilla/server/contexts/datasets.py:
```diff
@@ -619,7 +619,7 @@ async def create_records(
         await db.flush(records)
-        await _load_users_from_record_responses(records)
+        await refresh_records(records)
         await search_engine.index_records(dataset, records)
```
Not sure if it's the best approach but to explicitly link the preload of the records with the search engine maybe we can do the following:

```suggestion
        await search_engine.index_records(
            dataset, 
            await _preload_records_associations(records),
        )
```

Or if you don't want to do that:

```suggestion
        await _preload_records_associations(records)
        await search_engine.index_records(dataset, records)
```

jfcalvo commented on src/argilla/server/contexts/datasets.py:
```diff
@@ -866,6 +879,8 @@ async def update_records(
             await dataset.awaitable_attrs.vectors_settings
+            # This must be done to be sure that the engine index the updated vectors property
+            await refresh_records(records)
             await search_engine.index_records(dataset, records)
```
As said before we can do:

```suggestion
        await search_engine.index_records(
            dataset, 
            await _preload_records_associations(records)
        )
```

If you don't want to do that in any case you can change it to:


```suggestion
            await _preload_records_associations(records)
            await search_engine.index_records(dataset, records)
```

jfcalvo commented on src/argilla/server/contexts/datasets.py:
```diff
@@ -905,6 +920,7 @@ async def update_record(
         if needs_search_engine_update:
             await record.dataset.awaitable_attrs.vectors_settings
+            await refresh_records([record])
             await search_engine.index_records(record.dataset, [record])
```
As said before:

```suggestion
            await search_engine.index_records(
                record.dataset,
                [await _preload_record_associations(record)]
            )
```

Or:

```suggestion
            await _preload_record_associations(record)
            await search_engine.index_records(record.dataset, [record])
```

frascuchon replied:
For the first approach, that new function should return the loaded record list. We can review it together

jfcalvo commented on src/argilla/server/search_engine/commons.py:
```diff
@@ -58,72 +55,12 @@
 ALL_RESPONSES_STATUSES_FIELD = "all_responses_statuses"
 
 
-class SearchDocumentGetter(GetterDict):
```
Loving these code deletions.

jfcalvo commented on src/argilla/server/search_engine/commons.py:
```diff
@@ -211,7 +152,38 @@ def es_mapping_for_metadata_property(metadata_property: MetadataProperty) -> dic
+        raise Exception(f"Index configuration for metadata property of type {property_type} cannot be generated")
+
+
+def es_mapping_for_question_type(question_type: QuestionType):
```
Maybe I'm missing something (tell me if I'm wrong) but should not be this function better named `es_mapping_for_question` and receive a question instead of the type?

```suggestion
def es_mapping_for_question(question: Question):
```

Inside you can do the `if question.type` conditional.

frascuchon committed:
tests: Unify search engine tests into a single test suite (#4318)

This PR changes and unifies tests related to the search engine.

- All tests which will return the same results for the different engine
implementations are placed in `test_commons.py`
- Specific tests for elasticsearch implementation are placed in
`test_elasticsearch.py`
- Specific tests for opensearch implementation are placed in
`test_elasticsearch.py`

Refs: https://github.com/argilla-io/argilla/pull/4318

---------

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

frascuchon committed:
chore: Apply suggestion

frascuchon committed:
Update src/argilla/server/contexts/datasets.py

Co-authored-by: José Francisco Calvo <jose@argilla.io>

frascuchon committed:
Merge branch 'feat/index-records-with-suggestions-for-search-engine' of github.com:argilla-io/argilla into feat/index-records-with-suggestions-for-search-engine

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

frascuchon committed:
chore: Apply suggestion

frascuchon committed:
Merge branch 'feat/index-records-with-suggestions-for-search-engine' of github.com:argilla-io/argilla into feat/index-records-with-suggestions-for-search-engine

frascuchon committed:
fix: missing function renaming

gabrielmbmb commented on src/argilla/server/contexts/datasets.py:
```diff
@@ -788,6 +788,24 @@ async def _update_record(
+    await record.awaitable_attrs.vectors
+
+    for response in record.responses:
+        await response.awiatable_attrs.user
```
```suggestion
        await response.awaitable_attrs.user
```

gabrielmbmb commented on src/argilla/server/contexts/datasets.py:
```diff
@@ -788,6 +788,24 @@ async def _update_record(
+        await response.awiatable_attrs.user
+
+    for suggestion in record.suggestions:
+        await suggestion.awaitable_attrs.question
```
I think that here we could do `await record.dataset.awaitable_attrs.questions`

gabrielmbmb commented on src/argilla/server/search_engine/commons.py:
```diff
@@ -500,6 +469,64 @@ def _build_response_status_filter(status_filter: UserResponseStatusFilter) -> Di
     def _inverse_vector(self, vector_value: List[float]) -> List[float]:
         return [vector_value[i] * -1 for i in range(0, len(vector_value))]
 
+    def _map_record_to_es_document(self, record: Record) -> Dict[str, Any]:
```
👌🏻

gabrielmbmb commented on src/argilla/server/contexts/datasets.py:
```diff
@@ -788,6 +788,24 @@ async def _update_record(
+        await _preload_record_associations(record)
+
+
+async def _preload_record_associations(record: Record) -> None:
```
ignore my comments, if at the end we're going to query the data explicitly

jfcalvo committed:
feat: improve how relationships are loaded from database before index to only do one query per record

jfcalvo committed:
fix: some failing tests
